cmd/shim/main.go

```diff
@@ -144,6 +145,31 @@ func run() int {
 			stderr = io.MultiWriter(f, os.Stderr)
 		}
 
+		// Mount in the engine-session binary so it's available for nested sessions, set up env
+		// vars to point to it.
+		engineSessionBinPath := "/usr/bin/dagger-engine-session-linux-" + runtime.GOARCH // TODO: const
+		spec.Mounts = append(spec.Mounts, specs.Mount{
+			Destination: "/.dagger-engine-session", // TODO: const
+			Type:        "bind",
+			Source:      engineSessionBinPath,
+			Options:     []string{"rbind", "ro"},
+		})
+		// TODO: don't override if already set
+		spec.Process.Env = append(spec.Process.Env, "DAGGER_HOST=bin:///.dagger-engine-session") // TODO: const
+
+		// Also mount in the runner socket so it's available for nested sessions, set up env too
+		runnerSocketPath := "/run/buildkit/buildkitd.sock" // TODO: const
```

Comment:
Not sure I follow -- we're not "proxying" buildkit anymore?

A few questions:

- Do we need to remove the proxy code in engine?
- What are the trade-offs? Especially around version compatibility, is it better like this or if we can drive what the extension/d-in-d "see" as buildkit?
  - I'm guessing it's better like this since the shim lives in the runner, it's versioned together, and the runner knows best how to talk to itself

Reply:
> I'm guessing it's better like this since the shim lives in the runner

... and I just realized my 2 comments are contradicting -- either this logic lives in the engine or it's a feature flag, can't have both right?

Reply:
> Not sure I follow -- we're not "proxying" buildkit anymore?

No, and that's a good thing I think. If you break down what was actually happening before when we proxied the buildkitd api over a socket it was kind of crazy:
1. The "outer" client (i.e. on an actual host, not dind) opens a session with buildkitd
2. The outer client attaches an ssh socket provider to its session that opens another client session to buildkitd
3. A unix socket is mounted into an exec in buildkitd
4. When an "inner" (dind) client connects to that unix socket in the exec the data goes:
   * To buildkitd
   * ***Forwarded back*** to the outer client (where the ssh socket session attachable is)
   * Then once again ***fowarded back to buildkitd*** (!)

So while that was fine to get going, it wasn't really ideal.

Our new ability to bind mount in stuff direct from the engine container lets us way simplify this though by just providing the socket directly. Similarly, the fact that we run the shim outside the container actually gives us a route to maybe not needing to bind mount it in at all (and instead proxy it through the shim), though I didn't implement that yet.

> What are the trade-offs? Especially around version compatibility, is it better like this or if we can drive what the extension/d-in-d "see" as buildkit? 

One thing, the tradeoffs here are the exact same as before. We were providing the same buildkitd API before as we are now, it's just the data doesn't need to stream back and forth all over the place.

But in general, the tradeoff is that, yeah, technically users can get directly at the buildkit api right now. I doubt that's going to be true forever, so someday taking this away would be a breaking change.
* However, the buildkit API **only needs to be directly exposed for our test cases** where we want each `dagger.Connect` to result in a new session
* In the much more common case where the whole exec can share one session, this doesn't need to be exposed at all

The above is why I was mentioning in the description that I'm trying to find an easy way to allow us to default to the "single session" behavior (whereas the current implementation here only enables the "multiple session" behavior, just because it was easier to implement).
* I think I may have found an easy way, will update the PR with it
* Then, the plan will be something like this:
   * In the short term, our test cases still need the "multiple session" behavior, so I have an approach for allowing that to still work which will be hidden (and ugly, but that's fine for the short term)
   * Then, I think there's a nice path to cleanly supporting that multiple sessions behavior by extracting out one chunk of @vito's PR for dynamic sessions. We won't need the whole thing, but we can re-use part of it I think to get this.

Either way, I'm making some updates and will summarize the end state when I push it, should help clarify this and the tradeoffs.